src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -1755,10 +1740,15 @@ fn compare_old_fingerprint(
     old_hash_path: &Path,
     new_fingerprint: &Fingerprint,
     mtime_on_use: bool,
-) -> CargoResult<Option<DirtyReason>> {
+    forced: bool,
+) -> Option<DirtyReason> {
     let compare = _compare_old_fingerprint(old_hash_path, new_fingerprint, mtime_on_use);
     log_compare(unit, &compare);
-    compare
+    match compare {
+        Ok(None) if forced => Some(DirtyReason::Forced),
+        Ok(reason) => reason,
+        Err(_) => Some(DirtyReason::FreshBuild),
```

Comment:
Should `FreshBuild` contain the error?

Reply:
This could be done in theory, but `anyhow::Error` doesn't implement `Clone`….